include/deal.II/base/derivative_form.h

```diff
@@ -285,8 +288,9 @@ operator Tensor<1, dim, Number>() const
 
 
 template <int order, int dim, int spacedim, typename Number>
+template <typename Number2>
 inline DerivativeForm<order, dim, spacedim, Number>::
-operator Tensor<order + 1, dim, Number>() const
+operator Tensor<order + 1, dim, Number2>() const
```

Comment:
This doesn't sound quite right to me. Shouldn't the operator just return a `Tensor` representation of what it stores internally, and then if someone wants this in another type they can later cast from that tensor to one with a different scalar?

Note that that's basically what you do in the implementation, which now looks as follows:
```
template <int order, int dim, int spacedim, typename Number>
template <typename Number2>
inline DerivativeForm<order, dim, spacedim, Number>::
operator Tensor<order + 1, dim, Number>() const
{
  Assert((dim == spacedim), ExcMessage("Only allowed when dim==spacedim."));

  Tensor<order + 1, dim, Number> t;       <<<<<<<<<<<<<<<<<< old tensor type
  if (dim == spacedim)
    for (unsigned int j = 0; j < dim; ++j)
      t[j] = (*this)[j];
  return t;                                        <<<<<<<<<<<< cast to Tensor<....,Number2> happens here
}
```
My preference would be to make the place that calls this function do the cast at the calling site, not internally here.

Reply:
I tend to agree with @bangerth here, since this only converts the end result, I would prefer doing that explicitly in `FEPointEvaluation` (I was confused by the abbreviation FPE, which I read as floating point exception).

Reply:
That the tensor used the old number- type was not indended:
```cpp
Tensor<order + 1, dim, Number> t;       <<<<<<<<<<<<<<<<<< old tensor type
```
I have adjusted that. 

Since the `operator()` is for casting, the casting is already done explicitly done in `FEPointEvaluation`:

https://github.com/dealii/dealii/blob/1195dbbee9e40b56a0f7f1b1f5dd12b904b45e49/include/deal.II/matrix_free/fe_point_evaluation.h#L950-L954

The alternative would to cast from `DerivativeForm` to `typename internal::FEPointEvaluation::EvaluatorTypeTraits<dim, n_components, double>::gradient_type` (please note that one has to use double instead of Number in this long expression) and after that cast it to a float version (or do an assignment, which seems to broke in `Tensor`).

Reply:
I think what I meant to say is this: I consider the cast operator as saying "give me a tensor representation of this object", the emphasis being on "representation of this object" rather than "give me". The emphasis here is on *the object*. As a consequence, a `DerivativeForm<...,double>` should return a `Tensor<...,double>`.

In contrast, your implementation is for the question "I want to use this object as an X", in which the emphasis is on *want to use* -- a caller side consideration. I would prefer an explicit cast to `Tensor<...,Number2>` in such situations.